File: 094/binary-tree-inorder-traversal.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None


class BinaryTreeInorderTraversal:

    # Recursive solution.
    # def inorderTraversal(self, root):
    #     """
    #     :type root: TreeNode
    #     :rtype: List[int]
    #     """
    #     results = []
    #     if root is None:
    #         return results
    #
    #     if root.left is not None:
    #         results.extend(self.inorderTraversal(root.left))
    #
    #     results.append(root.val)
    #
    #     if root.right is not None:
    #         results.extend(self.inorderTraversal(root.right))
    #
    #     return results

    # Iterative solution.
    def inorderTraversal(self, root):
        """
        :type root: TreeNode
        :rtype: List[int]
        """
        result = []
        if root is None:
            return result

        stack = []
        while root is not None or len(stack) != 0:
            while root is not None:
                stack.append(root)
                root = root.left
            root = stack[len(stack) - 1]
            del stack[len(stack) - 1]
            result.append(root.val)
            root = root.right

        return result


class TreeNode(object):
    def __init__(self, x):
        self.val = x
        self.left = None
        self.right = None

    def __eq__(self, other):
        return self.val == other.val and self.left == other.left and self.right == other.right

File: 094/binary-tree-inorder-traversal.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val) {
 *     this.val = val;
 *     this.left = this.right = null;
 * }
 */
/**
 * @param {TreeNode} root
 * @return {number[]}
 */
var inorderTraversal = function (root) {
    function dfs(root, ans) {
        if (!root) return;

        dfs(root.left, ans);
        ans.push(root.val);
        dfs(root.right, ans);
    }

    let ans = [];
    dfs(root, ans);
    return ans;
};